### DISCUSSION TITLE
Safeboot and a custom username

### DISCUSSION THREAD

**Discussion by vidilab** [arendst/Tasmota#18410] [Q&A]
We build firmware with a custom username (WEB_USERNAME) for the web-interface that is different from the default admin. When upgrading, the safeboot partition requires "admin" as username. When the new firmware boots, it requires our custom username. This repeats on each upgrade.
Can the safeboot partition somehow be replaced by another version that knows about the right username?
Or maybe another option would be to remove the safeboot partition and switch back to the old layout. But esptool seems to preserve the partition structure.

**Comment by barbudor** [comment]
You can build a customized version of safeboot the same way you did for your own firmware as it is part of the standard environment targets for platformio
Then you can upgrade the existing safeboot through OTA:
- Set OtaUrl to point to a webserver hosting your custom safeboot
- Use command `Upload 2` to force update of safeboot

You will end up un safe boot so you will need to normally re-upload then your own custom firmware to get it back

Where as to build you own factory image to flash both in a single shot, you may have to go in the deep details of platformio
There, on @Jason2866 can guide you
Unless it already that automatically ?

**Reply by sfromis** [reply]
The path/name of the safeboot binary will be based on the binary file name in `OtaUrl`, modified to insert or replace `-safeboot` between the first part, and `.bin`. Hence, you should not put your safeboot binary name directly in OtaUrl.

Note that in safeboot mode, you cannot modify settings, like setting another `OtaUrl`.

**Reply by barbudor** [reply]
Check the doc about 'Upload 2'

**Reply by barbudor** [reply]
Form the doc and what Jason told me, you have to put the safeboot URL in OtaUrl
If it is not the case, then doc should be clearer
> 2 = (ESP32 only) download safeboot firmware based on OtaUrl

I read that as "OtaURl should point to your safeboot binary before running Upload 2"

**Reply by sfromis** [reply]
I read "based on" as the OtaUrl being modified to fetch safeboot from the same server as the regular binary. I've done many safeboot upgrades based on this, when OtaUrl points to my own web server. Thus, the URL should point at the server, facilitating an upgrade flow with both safeboot and regular binary being served from the same location, without putting the specific safeboot name in `OtaUrl`.

While it may work to upgrade **to** safeboot with the URL directly (as the URL modification would replace safeboot by safeboot), I'm "less convinced" that the next step of upgrading to the regular binary could work via setting `OtaUrl` followed by `Upgrade 1`. Still, if you "insist", you could then switch out of safeboot mode, back to the regular binary by `http://tasmotadevice/u4?u4=ota`.

You could say that safeboot URL for ESP32 works similar to minimal for ESP8266, except that it does not have to be fetched every time.

**Reply by barbudor** [reply]
I never wrote 'upgrade 1'
So it surely work as you say but I find 'based on' as misleading

**Reply by sfromis** [reply]
It would be misleading to not state that it is *only* "based on", but obviously it could be clarified.

> I never wrote 'upgrade 1'

`Upgrade 1` is normally a good way to "re-upload then your own custom firmware", but indeed not the only one. My reason for adding that sentence was primarily to explain how an upgrade flow can work with **one** value in `OtaUrl`, without having to switch the value for upgrading the safeboot binary.

**Comment by Jason2866** [comment]
The build process for a custom safeboot version is straight forward. Delete all folders beginning with `framework*` in hidden folder `.platformio/packages`

To build:
- Compile first!! all safeboot firmware variants.
- Compile your Tasmota variants

Tasmota Platformio setup checks before flashing with esptool.py (used in platformio too) if a safeboot firmware exists.
If not it does download from official Tasmota OTA server. In the described workflow, you generate valid firmware safeboot variants. The check will find yours and use it.

Yes, the OTA URL you set in your Tasmota firmware is used as source for safeboot too

The safeboot firmwares are stored in folder `.platformio/packages/framework-arduino*/variants/tasmota/*`

**Comment by vidilab** [comment]
Thanks for all replies.

The procedure to first compile safeboot, and then normal firmware, works, and solved the username problem. Because our firmware filename is non-standard, I had to rename the resulting safeboot file (in ``framework-arduinoespressif32\variants\tasmota`` from ``CUSTOM-safeboot.bin`` to ``tasmota32c3-safeboot.bin``) to make it work, but then it compiled. It can be flashed with esptool successfully.

Our OTA URL is not pointing to a file but a script with URL parameters, I will have to check how the automatic URL -safeboot modification will work out.

As alternative, because in our situation upgrading of safeboot part might be needed, and can be difficult for the user, I am also considering disabling the safeboot scheme, using for example (in platform_tasmota_env.ini):
```
board_build.partitions  = partitions/esp32_partition_app1856k_fs320k.csv
```
After flashing with esptool, this reverts the device back to the old partition scheme with ``app0`` and ``app1``.

**Comment by Jason2866** [comment]
Two partitions (old scheme) is an option too. See the example in `platformio_tasmota_cenv_sample.ini`

```
[env:tasmota32s3-file]
extends                     = env:tasmota32_base
board                       = esp32s3-qio_qspi
board_build.f_cpu           = 240000000L
board_build.f_flash         = 80000000L
build_flags                 = ${env:tasmota32_base.build_flags}
                              -DFIRMWARE_TASMOTA32
                              -DOTA_URL='"http://ota.tasmota.com/tasmota32/release/tasmota32s3.bin"'
; !!! Real flash size needed, avoid autoresize since it is formating FS !!!
board_upload.flash_size     = 8MB
board_upload.maximum_size   = 8388608
; Without autoresize a partition scheme is needed which does fit to flash size
board_build.partitions      = partitions/esp32_partition_app2944k_fs2M.csv
; Dont use safeboot, not used in this partition scheme -> an empty entry needed to overwrite the default setting
board_upload.arduino.flash_extra_images =
; Example for custom file upload in Tasmota Filesystem
custom_files_upload         = ${env:tasmota32_base.custom_files_upload}
                              tasmota/berry/modules/Partition_Wizard.tapp
                              https://github.com/tasmota/autoconf/raw/main/esp32s3/DevKitC-1.autoconf
```
